File: midolman/src/main/scala/org/midonet/midolman/simulation/ICMP.scala
/*
 * Copyright 2016 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.midolman.simulation

import org.midonet.odp.FlowMatch
import org.midonet.packets.ICMP.{EXCEEDED_CODE, UNREACH_CODE}
import org.midonet.packets._

object Icmp {

    trait IcmpErrorSender[IP <: IPAddr] {
        /**
         * Send an ICMP error message.
         */
        protected def icmpAnswer(inPort: RouterPort,
                                 context: PacketContext,
                                 fmatch: FlowMatch,
                                 icmpType: Byte,
                                 icmpCode: Any): Option[Ethernet]

        /**
         * Will be called whenever an ICMP unreachable is needed for the given
         * IP version.
         */
        def unreachableProhibitedIcmp(inPort: RouterPort, context: PacketContext,
                                      fmatch: FlowMatch) =
            icmpAnswer(inPort, context, fmatch, ICMP.TYPE_UNREACH, UNREACH_CODE.UNREACH_FILTER_PROHIB)

        /**
         * Will be called whenever an ICMP Unreachable network is needed for the
         * given IP version.
         */
        def unreachableNetIcmp(inPort: RouterPort, context: PacketContext, fmatch: FlowMatch) =
            icmpAnswer(inPort, context, fmatch, ICMP.TYPE_UNREACH, UNREACH_CODE.UNREACH_NET)

        /**
         * Will be called whenever an ICMP Unreachable host is needed for the
         * given IP version.
         */
        def unreachableHostIcmp(inPort: RouterPort, context: PacketContext, fmatch: FlowMatch) =
            icmpAnswer(inPort, context, fmatch, ICMP.TYPE_UNREACH, UNREACH_CODE.UNREACH_HOST)

        def unreachableFragNeededIcmp(inPort: RouterPort, context: PacketContext, fmatch: FlowMatch) =
            icmpAnswer(inPort, context, fmatch, ICMP.TYPE_UNREACH, UNREACH_CODE.UNREACH_FRAG_NEEDED)

        /**
         * Will be called whenever an ICMP Time Exceeded is needed for the given
         * IP version.
         */
        def timeExceededIcmp(inPort: RouterPort, context: PacketContext, fmatch: FlowMatch) =
            icmpAnswer(inPort, context, fmatch, ICMP.TYPE_TIME_EXCEEDED, EXCEEDED_CODE.EXCEEDED_TTL)
    }

    implicit object IPv4Icmp extends IcmpErrorSender[IPv4Addr] {
        val invalidAddr = IPv4Addr(0xffffffff)

        override def icmpAnswer(
                inPort: RouterPort,
                context: PacketContext,
                fmatch: FlowMatch,
                icmpType: Byte,
                icmpCode: Any): Option[Ethernet] = {
            context.log.debug("Prepare an ICMP response")
            val packet = context.ethernet

            // Check whether the original packet is allowed to trigger ICMP.
            if (inPort == null) {
                context.log.debug("Cannot send ICMP because input port is null")
                return None
            }
            if (inPort.portAddress4 eq null) {
                context.log.debug("Port {} does not have an IPv4 address",
                                  inPort.id)
                return None
            }
            if (!canSend(fmatch, inPort, context)) {
                context.log.debug("ICMP not allowed for this packet")
                return None
            }
            // Build the ICMP packet from inside-out: ICMP, IPv4, Ethernet headers.
            context.log.debug(s"Generating ICMP error $icmpType:$icmpCode")
            val icmp = buildError(icmpType, icmpCode, fmatch, packet)

            val ip = new IPv4()
            ip.setPayload(icmp)
            ip.setProtocol(ICMP.PROTOCOL_NUMBER)
            // The nwDst is the source of triggering IPv4 as seen by this router.
            ip.setDestinationAddress(fmatch.getNetworkSrcIP.asInstanceOf[IPv4Addr])
            // The nwSrc is the address of the ingress port.
            ip.setSourceAddress(inPort.portAddress4.getAddress)
            val eth = new Ethernet()
            eth.setPayload(ip)
            eth.setEtherType(IPv4.ETHERTYPE)
            eth.setSourceMACAddress(inPort.portMac)
            eth.setDestinationMACAddress(fmatch.getEthSrc)

            Some(eth)
        }

        private def buildError(icmpType: Byte, icmpCode: Any,
                               forMatch: FlowMatch, forPacket: Ethernet)
                : ICMP = {
            forMatch.applyTo(forPacket)
            val ipPkt = forPacket.getPayload.asInstanceOf[IPv4]
            icmpCode match {
                case c: ICMP.EXCEEDED_CODE if icmpType == ICMP.TYPE_TIME_EXCEEDED =>
                    val icmp = new ICMP()
                    icmp.setTimeExceeded(c, ipPkt)
                    icmp
                case c: ICMP.UNREACH_CODE if icmpType == ICMP.TYPE_UNREACH =>
                    val icmp = new ICMP()
                    icmp.setUnreachable(c, ipPkt)
                    icmp
                case _ =>
                    null
            }
        }

        /**
         * Determine whether a packet can trigger an ICMP error.  Per RFC 1812 sec.
         * 4.3.2.7, some packets should not trigger ICMP errors:
         *   1) Other ICMP errors.
         *   2) Invalid IP packets.
         *   3) Destined to IP bcast or mcast address.
         *   4) Destined to a link-layer bcast or mcast.
         *   5) With source network prefix zero or invalid source.
         *   6) Second and later IP fragments.
         *
         * @return True if-and-only-if the packet meets none of the above conditions
         *         - i.e. it can trigger an ICMP error message.
         */
        private def canSend(fmatch: FlowMatch, outPort: RouterPort,
                            context: PacketContext): Boolean = {
            if (fmatch.getEtherType != IPv4.ETHERTYPE)
                return false

            if (fmatch.getNetworkProto == ICMP.PROTOCOL_NUMBER &&
                ICMP.isError(fmatch.getSrcPort.toByte)) {
                    context.log.debug("Skipping generation of ICMP error for ICMP error packet")
                    return false
            }

            // TODO(pino): check the IP packet's validity - RFC1812 sec. 5.2.2
            // Ignore packets to IP mcast addresses.
            if (fmatch.getNetworkDstIP.isMcast) {
                context.log.debug("Not generating ICMP Unreachable for packet "+
                                  "to an IP multicast address.")
                return false
            }
            // Ignore packets sent to the local-subnet IP broadcast address of the
            // intended egress port.
            if (null != outPort && (outPort.portAddress4 ne null) &&
                fmatch.getNetworkDstIP == outPort.portAddress4.toBroadcastAddress) {
                context.log.debug("Not generating ICMP Unreachable for packet to "
                                + "the subnet local broadcast address.")
                return false
            }
            // Ignore packets to Ethernet broadcast and multicast addresses.
            if (fmatch.getEthDst.mcast) {
                context.log.debug("Not generating ICMP Unreachable for packet to "
                                + "Ethernet broadcast or multicast address.")
                return false
            }
            // Ignore packets with source network prefix zero or invalid source.
            // TODO(pino): See RFC 1812 sec. 5.3.7
            if (fmatch.getNetworkSrcIP == invalidAddr ||
                fmatch.getNetworkDstIP == invalidAddr) {
                context.log.debug(
                    "Not generating ICMP Unreachable for all-hosts broadcast packet")
                return false
            }
            // TODO(pino): check this fragment offset
            // Ignore datagram fragments other than the first one.
            if (fmatch.getIpFragmentType == IPFragmentType.Later) {
                context.log.debug(
                    "Not generating ICMP Unreachable for IP fragment packet")
                return false
            }

            true
        }
    }
}

File: midonet-util/src/main/java/org/midonet/packets/ICMP.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.packets;

import java.nio.ByteBuffer;
import java.util.Arrays;

public class ICMP extends BasePacket {
    public static final byte PROTOCOL_NUMBER = 1;

    public static final byte CODE_NONE = 0;

    public static final byte TYPE_UNREACH = 3;
    public static enum UNREACH_CODE {
        UNREACH_NET(0),
        UNREACH_HOST(1),
        UNREACH_PROTOCOL(2),
        UNREACH_PORT(3),
        UNREACH_FRAG_NEEDED(4),
        UNREACH_SOURCE_ROUTE(5),
        UNREACH_FILTER_PROHIB(13);

        private final byte value;
        private UNREACH_CODE(int value) { this.value = (byte)value; }
        public byte toByte() { return value; }
    }

    public static final byte TYPE_TIME_EXCEEDED = 11;
    public static enum EXCEEDED_CODE {
        EXCEEDED_TTL(0), EXCEEDED_REASSEMBLY(1);

        private final byte value;
        private EXCEEDED_CODE(int value) { this.value = (byte)value; }
        public byte toByte() { return value; }
    }

    public static final byte TYPE_ECHO_REPLY = 0;
    public static final byte TYPE_ECHO_REQUEST = 8;
    public static final byte TYPE_SOURCE_QUENCH = 4;
    public static final byte TYPE_REDIRECT = 5;
    public static final byte TYPE_ROUTER_SOLICITATION= 10;
    public static final byte TYPE_PARAMETER_PROBLEM = 12;

    /**
     * ICMP header length as a number of octets.
     */
    public static final int HEADER_LEN = 8;

    // Types

    private byte type;
    private byte code;
    private short checksum;
    private int quench;
    private byte[] data;

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("ICMP [type=").append((int)type);
        sb.append(", code=").append((int)code);
        sb.append(", cksum=").append(checksum);
        sb.append(", quench=").append(quench);
        sb.append("]");
        return sb.toString();
    }

    public static boolean isError(byte type) {
        return type != TYPE_ECHO_REQUEST && type != TYPE_ECHO_REPLY;
    }

    public boolean isError() {
        return isError(type);
    }

    public byte getType() {
        return type;
    }

    public byte getCode() {
        return code;
    }

    public short getChecksum() {
        return checksum;
    }

    public int getQuench() {
        return quench;
    }

    public final byte[] getData() {
        return data;
    }

    private void setIPv4Packet(IPv4 ipPkt) {
        IPacket payload = ipPkt.payload;
        while (!(payload instanceof Data) && payload != null) {
            payload = payload.getPayload();
        }
        int payloadLength = payload != null ? ((Data) payload).getData().length : 0;

        byte[] data = ipPkt.serialize();
        int length = ipPkt.totalLength - payloadLength;
        if (length >= data.length)
            this.data = data;
        else
            this.data = Arrays.copyOf(data, length);
    }

    public void setType(byte type, byte code, byte[] data) {
        this.type = type;
        this.code = code;
        this.data = (data == null) ? null : Arrays.copyOf(data, data.length);
    }

    public void setUnreachable(UNREACH_CODE unreachCode, IPv4 ipPkt) {
        type = TYPE_UNREACH;
        code = unreachCode.value;
        checksum = 0;
        quench = 0;
        setIPv4Packet(ipPkt);
    }

    public void setFragNeeded(int size, IPv4 ipPkt) {
        setUnreachable(UNREACH_CODE.UNREACH_FRAG_NEEDED, ipPkt);
        quench = size & 0xffff;
    }

    public void setTimeExceeded(EXCEEDED_CODE timeCode, IPv4 ipPkt) {
        type = TYPE_TIME_EXCEEDED;
        code = timeCode.value;
        checksum = 0;
        quench = 0;
        setIPv4Packet(ipPkt);
    }

    /**
     * BEWARE: direct assignment of data
     *
     * @param id
     * @param seq
     * @param data
     */
    public void setEchoRequest(short id, short seq, byte[] data) {
        type = TYPE_ECHO_REQUEST;
        code = CODE_NONE;
        checksum = 0;
        quench = id << 16 | seq & 0xffff;
        this.data = data;
    }

    /**
     * BEWARE: direct assignment of data
     *
     * @param id
     * @param seq
     * @param data
     */
    public void setEchoReply(short id, short seq, byte[] data) {
        type = TYPE_ECHO_REPLY;
        code = CODE_NONE;
        checksum = 0;
        quench = (id << 16) | (seq & 0xffff);
        this.data = data;
    }

    public void setIdentifier(short id) {
        quench |= id << 16;
    }

    public short getIdentifier() {
        return (short)(quench >> 16);
    }

    public short getSequenceNum() {
        return (short)quench;
    }

    /**
     * BEWARE: direct assignment of data
     *
     * @param data
     */
    public void setData (byte[] data) {
        this.data = data;
    }

    public void clearChecksum() {
        checksum = 0;
    }

    /**
     * Serializes the packet. Will compute and set the checksum if it's set to
     * 0 (zero) at the time serialize is called.
     */
    @Override
    public byte[] serialize() {
        int length = HEADER_LEN + ((data == null) ? 0 : data.length);
        byte[] bytes = new byte[length];
        ByteBuffer bb = ByteBuffer.wrap(bytes);
        bb.put(type);
        bb.put(code);
        bb.putShort(checksum);
        bb.putInt(quench);

        if (null != data)
            bb.put(data);

        // ICMP checksum is calculated on header+data, with checksum=0.
        if (checksum == 0) {
            checksum = IPv4.computeChecksum(bytes, 0, bb.position(), 2);
            bb.putShort(2, this.checksum);
        }
        return bytes;
    }

    @Override
    public IPacket deserialize(ByteBuffer bb) throws MalformedPacketException {

        // Check that the size is correct to avoid BufferUnderflowException.
        if (bb.remaining() < HEADER_LEN) {
            throw new MalformedPacketException("Invalid ICMP header len: "
                    + bb.remaining());
        }

        type = bb.get();
        code = bb.get();
        checksum = bb.getShort();
        quench = bb.getInt();
        if (bb.hasRemaining()) {
            this.data = new byte[bb.remaining()];
            bb.get(this.data);
        }
        return this;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (!super.equals(obj))
            return false;
        if (!(obj instanceof ICMP))
            return false;
        final ICMP other = (ICMP) obj;
        if (this.type != other.type)
            return false;
        if (this.code != other.code)
            return false;
        if (this.checksum != other.checksum)
            return false;
        if (this.quench != other.quench)
            return false;
        return java.util.Arrays.equals(this.data, other.data);
    }
}
